### PR TITLE
[draft][WS1][kernels] Batch-invariant logprob Ascend C Kernel

### PR DESCRIPTION
**Pull request by erfgss**
## Summary

Adds an Ascend C (CANN) backend for the batch-invariant selected-token
log-probability operator introduced in #199 and extended to SM90 CUDA in #204.
The operator computes, from **already materialized** logits:
```text
logp[t] = logits[t, target_ids[t]] - logsumexp(logits[t, :])
```

The kernel streams each vocab row through UB in fixed 4096-element tiles with a
two-pass (max, then sum-exp) fixed-order fp32 reduction, so the per-row result
depends only on the row itself — never on batch size or layout. Each row is
processed end-to-end by exactly one AI core block, and outputs are staged in UB
and written with `DataCopyPad` (scalar GM `SetValue`/`GetValue` are avoided per
known hardware pitfalls). It is registered as the top-priority backend on the
new `npu` platform key and falls back cleanly to the PyTorch native op.

| Path | Status |
| --- | --- |
| Ascend C forward (`csrc/ascend/batch_invariant_logp_ascend.asc`) | Two-pass streaming reduction, bf16 + fp32, one block per row. New. |
| Python op (`BatchInvariantLogpAscendOp`) | Autograd wrapper; Ascend C forward + PyTorch-formula backward reusing the forward-saved `lse`. New. |
| Extension (`rl_engine._C_npu`) | New bisheng-built (`-x asc`) extension, gated on `KERNEL_ALIGN_FORCE_ASCEND=1`; `--npu-arch` defaults to `dav-2201`, override via `KERNEL_ALIGN_ASCEND_ARCH`. New. |
| Platform | `DeviceContext` now detects `torch_npu` / `torch.npu`; registry gains the `"npu"` platform key. New. |
| Registry | `batch_invariant_logp` on NPU: Ascend -> PyTorch. Ascend instantiation failure (extension not built) degrades to native. |
| Fallback | Non-bf16/fp32 or non-NPU tensors -> PyTorch native (Triton rejects non-CUDA devices). |
| Benchmark | `benchmarks/benchmark_batch_invariant_logp.py` dispatches timing/VRAM through `torch.npu` and adds an `ascend` column. |

## Implementation

- **New kernel + binding**: `csrc/ascend/batch_invariant_logp_ascend.asc`
  (kernel class, `__global__ __vector__` entries, host entrypoint, pybind).
- **Build**: `setup.py` (`AscendBuildExtension` invoking `bisheng`, following
  the official torch_npu `cpp_extension_asc` pattern), `envs.py`
  (`KERNEL_ALIGN_FORCE_ASCEND`, `KERNEL_ALIGN_ASCEND_ARCH`), `rl_engine/_C_npu.pyi`.
- **Python op**: `rl_engine/kernels/ops/ascend/loss/batch_invariant_logp.py`,
  interface aligned with `BatchInvariantLogpSM90Op`.
- **Platform/registry**: `rl_engine/platforms/device.py`,
  `rl_engine/kernels/registry.py`.
- **Tests/docs/benchmark**: `tests/test_batch_invariant_logp.py`,
  `rl_engine/kernels/gtest/operator_specs.py`,
  `docs/operators/batch-invariant-logp.md`,
  `benchmarks/benchmark_batch_invariant_logp.py`.

## Validation environment

| Item | Value |
| --- | --- |
| NPU | Ascend (dav_c220, CANN dav-2201 target) |
| CANN toolkit | 9.1.0 |
| PyTorch / torch_npu | 2.7.1 / matching torch_npu |
| Python | 3.11 (aarch64) |
| Build | `KERNEL_ALIGN_FORCE_ASCEND=1 pip install -e . --no-build-isolation` |

## Correctness / Tests

```bash
python -m pytest tests/test_batch_invariant_logp.py -k Ascend -q -rs
```

- `TestAscendCorrectness` (fp32/bf16 vs `log_softmax + gather`, large vocab,
  single-token, 3-D, vs native op): **6/6 passed** on device.
- Intra-core pipeline sync uses per-pipe `SetFlag`/`WaitFlag` events;
  `AscendC::SyncAll()` (cross-core barrier) is deliberately avoided because it
  deadlocks when launched blocks exceed resident cores.

**WIP**: `TestAscendBatchInvariance` / `TestAscendBackward` /
`TestAscendIgnoreIndex` / `TestAscendFallback` are included but still being
re-verified on device after the sync fix; will update before undrafting.

CPU-only regression (no NPU): `tests/test_batch_invariant_logp.py`,
`tests/test_kernel_registry.py`, `rl_engine/tests/test_dispatch.py`,
`tests/test_cpu_hal.py` — all pass; Ascend tests skip cleanly.

## Files

- `csrc/ascend/batch_invariant_logp_ascend.asc` (new)
- `rl_engine/kernels/ops/ascend/loss/batch_invariant_logp.py` (new)
- `rl_engine/_C_npu.pyi` (new)
- `csrc/ascend/`, `rl_engine/kernels/ops/ascend/` package init files (new)
- `setup.py`, `envs.py`, `rl_engine/kernels/registry.py`,
  `rl_engine/platforms/device.py`
- `tests/test_batch_invariant_logp.py`,
  `rl_engine/kernels/gtest/operator_specs.py`,
  `docs/operators/batch-invariant-logp.md`,
  `benchmarks/benchmark_batch_invariant_logp.py`
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added Ascend NPU support for batch-invariant selected-token log probabilities.
  * Supports BF16 and FP32 inputs, gradient computation, ignored targets, and automatic fallback for unsupported configurations.
  * Added Ascend device detection, build configuration, and backend dispatch.

* **Performance**
  * Added accelerator-aware benchmarking across CUDA, ROCm, and Ascend, including latency, speedup, and memory reporting.

* **Documentation**
  * Documented Ascend support, configuration, dispatch behavior, supported data types, and fallback rules.

* **Tests**
  * Added Ascend coverage for correctness, batching, gradients, ignored targets, and fallback behavior.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: rl_engine/kernels/ops/ascend/loss/batch_invariant_logp.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2026 RL-Kernel Contributors
+
+from __future__ import annotations
+
+import torch
+
```
Comment (zhangj1an) [inline rl_engine/kernels/ops/ascend/loss/batch_invariant_logp.py:9]: Can you add these two lines to the file, 
`from typing import Any`
and 
`_C_npu = None → _C_npu: Any = None`

to pass a CI check called mypy.

You can check the CI lists and fix them if they are not green marks yet

File: rl_engine/kernels/gtest/operator_specs.py [hunk 2]
```diff
@@ -171,6 +171,8 @@ def _load_object(path: str) -> Any:
             "TritonBatchInvariantLogpOp",
             "cuda-sm90": "rl_engine.kernels.ops.cuda.loss.batch_invariant_logp."
             "BatchInvariantLogpSM90Op",
+            "ascend": "rl_engine.kernels.ops.ascend.loss.batch_invariant_logp."
+            "BatchInvariantLogpAscendOp",
```
Comment (zhangj1an) [inline rl_engine/kernels/gtest/operator_specs.py:191]: In order to let the new logprob kernel pass gtest (it is our in-house test to measure precision of a single-card kernel), you need to update the `scripts/check_operator.py` file with the following 3 content, 

inside the _select_device method, you need to add these two lines, 
```
if value == "npu" and not _npu_available():
        raise RuntimeError("--device npu was requested, but no Ascend NPU is available")
```

inside the parse_args method, you need to add the npu option,

```
parser.add_argument(
        "--candidate",
        default="pytorch",
        help="Candidate backend to validate, for example pytorch, cuda, cuda-sm90, triton, "
        "ascend.",
    )
```

i think you also need to add a new method, 

```
def _npu_available() -> bool:
    """torch.npu only exists after torch_npu is imported; probe defensively."""
    try:
        import torch_npu  # noqa: F401
    except ImportError:
        return False
    return hasattr(torch, "npu") and torch.npu.is_available()
```

**Review by zhangj1an — APPROVED** [review]
LGTM. I have tested this new kernel on NPU and it works. Please fix CI conflicts, thanks for your contribution!

You can use this command,
```
python scripts/check_operator.py --op batch_invariant_logp --candidate ascend \
      --device npu --dtype fp32 --batch 4 --seq 8 --vocab 2048 --check-grad
```

And then the ouptut for fp32 is 
```
INFO 08-13 21:34:55 [RL-Kernel]: Successfully linked to precompiled _C_npu.batch_invariant_logp_ascend kernel.
  /home/ma-user/work/z84450661/rl-kernel/rl_engine/kernels/ops/pytorch/loss/batch_invariant_logp.py:105: UserWarning: Cannot create tensor with interal format while
  allow_internel_format=False, tensor will be created with base format. (Triggered internally at ../torch_npu/csrc/aten/common/TensorFactories.cpp:340.)
    selected_logp = selected_logp.where(valid_mask, torch.zeros_like(selected_logp))
  suite=batch_invariant_logp passed=True pass_rate=1.0000
  candidate=ascend-batch_invariant_logp backend=ascend passed=True pass_rate=1.0000
    case=batch_invariant_logp-torch.float32-4x8x2048 output=0 shape=(4, 8) dtype=torch.float32 max_abs=9.53674316e-07 mean_abs=1.78813934e-07 max_rel=1.45964989e-07
  tol=(atol=1.000e-05, rtol=0.000e+00) passed=True
    case=batch_invariant_logp-torch.float32-4x8x2048 output=1 gradient:logits shape=(4, 8, 2048) dtype=torch.float32 max_abs=1.32247806e-07 mean_abs=1.55981117e-10
  max_rel=1.23174477e-05 tol=(atol=1.000e-05, rtol=0.000e+00) passed=True
```

The output for bf16 is 
```
 INFO 08-13 21:35:49 [RL-Kernel]: Successfully linked to precompiled _C_npu.batch_invariant_logp_ascend kernel.
  ...（同样的 UserWarning 两行）...
  suite=batch_invariant_logp passed=True pass_rate=1.0000
  candidate=ascend-batch_invariant_logp backend=ascend passed=True pass_rate=1.0000
    case=batch_invariant_logp-torch.bfloat16-4x8x2048 output=0 shape=(4, 8) dtype=torch.float32 max_abs=9.53674316e-07 mean_abs=2.98023224e-08 max_rel=1.10753255e-07
  tol=(atol=5.000e-02, rtol=0.000e+00) passed=True
```

So that means, 
forward:  max_abs=9.54e-07  tol=1e-05  ✓
gradient: max_abs=1.32e-07  tol=1e-05  ✓
bf16: passed=True（max_abs=9.54e-07, tol=5e-02）✓

Also the 29 unit test passed in `tests/test_op_checks.py`.

**Comment by Flink-ddd** [comment]
please resolve the CI error, other LGTM.

**Review by Flink-ddd — APPROVED** [review]
LGTM, Thank you for work!
